Close stale pr

apelisse opened the pull request:
Close pull-request that have not been modified in a while

Fixes #481

Commit:
Add new interface to query PullRequest comments

The current github interface only allows to retrieve comments for the
issue (that's the comment posted on the pull-request or the issue
directly).

Let's create a way to retrieve comment posted on the diff. Note that
this doesn't retrieve comments from the issue (must be fetched
separately).

eparis commented on mungegithub/github/github.go:
```diff
@@ -1289,6 +1292,41 @@ func (obj *MungeObject) MergePR(who string) error {
 }
 
 // ListComments returns all comments for the issue/PR in question
+func (obj *MungeObject) ListPRComments() ([]github.PullRequestComment, error) {
```
You should just use ListComments I think, no?

apelisse replied:
Unfortunately, ListComments doesn't return comments made on the diff of the pull-request, like this one.

eparis replied:
ewwww, does ListPRComments() give all of the issue comments? uggh..

apelisse replied:
OK, I'm going to change my names a little bit. There are two types of comments:
- Issue comments
- Review comments: these are related to a specific line of code in the diff

eparis commented on mungegithub/mungers/close-stale-pr.go:
```diff
@@ -0,0 +1,138 @@
+		if *comment.User.Login == botName || *comment.User.Login == jenkinsBotName {
+			continue
+		}
+		last_human = comment.UpdatedAt
```
are comments guaranteed in chronological order? Should you check time.After?

apelisse replied:
Good call, they're not. They're actually sorted by Id, which is usually going to lead to a chronological order ... unless you update an older comment.

eparis commented on mungegithub/mungers/close-stale-pr.go:
```diff
@@ -0,0 +1,138 @@
+	return last_human, nil
+}
+
+func lastHumanUpdateIssue(obj *github.MungeObject) (*time.Time, error) {
```
Why do I need both of these?

eparis commented:
I like the idea, but I feel like a 30 and 60 day warning would be nice...

apelisse commented:
OK I've updated the pull-request.

I have not implemented the warning yet, but will think about it.

I fixed:
- Comments that may not be chronologically sorted
- Hopefully, errors detected by `verify_all.sh`
- I've renamed PR to Review to use the same words as the github api documentation

eparis commented:
I'll merge but I think we should update the comment a little.  Just re-opening will get it immediately re-closed...

I think we should say how long it has been idle and we should say that adding a comment will restart the clock.

Can we also restart the clock on git push?  I'm not sure what we have available there...

apelisse commented:
Please feel free to wait before you merge that PR. There is no rush :-)
You have very good points, I'll try to see if they can be addressed

apelisse commented:
OK, I think this can be improved by tracking [events](https://developer.github.com/v3/issues/events/). 

Here is what we can do based on the data in events:
- reset the timer if the pull-request has been updated (new head commit id)
- reset the timer when the pull-request is reopened

And also:
- Change the comment to be more explicit ("Closing PR because no action in X days")
- Remove this comment whenever the pull-request is reopened

And I also figured out a way to add warnings, we'll talk about that later

Commit:
Warn before closing stale pull-request

Create a warning mechanism so that we don't close pull-requests
overnight. You can choose how much time before you want to be warned,
and how often you want to be reminded.

This should also remove previous warning comments from the list of
comments so that we don't clutter the pull-request log.

Commit:
Close inactive pull-request

Close a pull-request if it didn't have any human interaction in a given
time. It does so by going through all the comments of the
issue/pull-request not made by a bot and through events looking for a
re-open.

We can not find events such as a new push to the pull-request for the
moment.

The closing comment will be removed if the pull-request is re-opened

Commit:
Keep PR open if it has keep-open label

Somebody may want to keep a pull-request open indefinitely. Provide this
feature by allowing people to mark pull-requests with "keep-open"
label.

apelisse commented:
OK Just updated and added warning and label to 'keep-open'. Also added mechanism to remove stale comments... I have not tested the part that removes the comment (because we don't have such comments ...) so I'm kind of unsure about the regex.

Here is the output of a `--dry-run --once` run:
```
I0509 16:32:35.927334   20663 github.go:1365] Commenting "This PR hasn't been active in 69 days. Will be closed in 20 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 13216
I0509 16:32:39.945247   20663 github.go:1365] Commenting "This PR hasn't been active in 83 days. Will be closed in 6 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 15385
I0509 16:32:40.784415   20663 github.go:1365] Commenting "This PR hasn't been active in 80 days. Will be closed in 9 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 15927
I0509 16:32:42.662823   20663 github.go:1365] Commenting "This PR hasn't been active in 54 days. Will be closed in 35 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 16146
I0509 16:32:43.604546   20663 github.go:1365] Commenting "This PR hasn't been active in 42 days. Will be closed in 47 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 16933
I0509 16:32:46.272063   20663 github.go:1365] Commenting "This PR hasn't been active in 82 days. Will be closed in 7 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 17085
I0509 16:32:46.972465   20663 github.go:1365] Commenting "This PR hasn't been active in 40 days. Will be closed in 49 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 17305
I0509 16:32:47.851478   20663 github.go:1365] Commenting "This PR hasn't been active in 108 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 17393
I0509 16:32:47.851529   20663 github.go:1146] Closing PR# 17393
I0509 16:32:48.546276   20663 github.go:1365] Commenting "This PR hasn't been active in 40 days. Will be closed in 49 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 17543
I0509 16:32:50.365445   20663 github.go:1365] Commenting "This PR hasn't been active in 34 days. Will be closed in 55 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 17688
I0509 16:32:51.967198   20663 github.go:1365] Commenting "This PR hasn't been active in 77 days. Will be closed in 12 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 17893
I0509 16:32:52.440090   20663 github.go:1365] Commenting "This PR hasn't been active in 95 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 17926
I0509 16:32:52.440139   20663 github.go:1146] Closing PR# 17926
I0509 16:32:55.258515   20663 github.go:1365] Commenting "This PR hasn't been active in 63 days. Will be closed in 26 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 18308
I0509 16:32:59.108488   20663 github.go:1365] Commenting "This PR hasn't been active in 111 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 19256
I0509 16:32:59.108540   20663 github.go:1146] Closing PR# 19256
I0509 16:33:00.123129   20663 github.go:1365] Commenting "This PR hasn't been active in 124 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 19283
I0509 16:33:00.123160   20663 github.go:1146] Closing PR# 19283
I0509 16:33:01.793335   20663 github.go:1365] Commenting "This PR hasn't been active in 121 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 19442
I0509 16:33:01.793379   20663 github.go:1146] Closing PR# 19442
I0509 16:33:02.369373   20663 github.go:1365] Commenting "This PR hasn't been active in 80 days. Will be closed in 9 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 19445
I0509 16:33:03.525708   20663 github.go:1365] Commenting "This PR hasn't been active in 37 days. Will be closed in 52 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 19473
I0509 16:33:04.061525   20663 github.go:1365] Commenting "This PR hasn't been active in 96 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 19485
I0509 16:33:04.061577   20663 github.go:1146] Closing PR# 19485
I0509 16:33:05.287798   20663 github.go:1365] Commenting "This PR hasn't been active in 110 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 19534
I0509 16:33:05.287850   20663 github.go:1146] Closing PR# 19534
I0509 16:33:05.900072   20663 github.go:1365] Commenting "This PR hasn't been active in 56 days. Will be closed in 33 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 19606
I0509 16:33:06.423984   20663 github.go:1365] Commenting "This PR hasn't been active in 69 days. Will be closed in 20 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 19625
I0509 16:33:06.933489   20663 github.go:1365] Commenting "This PR hasn't been active in 68 days. Will be closed in 21 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 19706
I0509 16:33:07.492733   20663 github.go:1365] Commenting "This PR hasn't been active in 94 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 19708
I0509 16:33:07.492788   20663 github.go:1146] Closing PR# 19708
I0509 16:33:07.952850   20663 github.go:1365] Commenting "This PR hasn't been active in 115 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 19710
I0509 16:33:07.952898   20663 github.go:1146] Closing PR# 19710
I0509 16:33:08.918394   20663 github.go:1365] Commenting "This PR hasn't been active in 95 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 19763
I0509 16:33:08.918447   20663 github.go:1146] Closing PR# 19763
I0509 16:33:09.440517   20663 github.go:1365] Commenting "This PR hasn't been active in 74 days. Will be closed in 15 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 19804
I0509 16:33:10.612886   20663 github.go:1365] Commenting "This PR hasn't been active in 81 days. Will be closed in 8 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 19872
I0509 16:33:11.126119   20663 github.go:1365] Commenting "This PR hasn't been active in 90 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 19919
I0509 16:33:11.126173   20663 github.go:1146] Closing PR# 19919
I0509 16:33:12.444373   20663 github.go:1365] Commenting "This PR hasn't been active in 55 days. Will be closed in 34 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 19951
I0509 16:33:12.941828   20663 github.go:1365] Commenting "This PR hasn't been active in 56 days. Will be closed in 33 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 19996
I0509 16:33:13.972335   20663 github.go:1365] Commenting "This PR hasn't been active in 88 days. Will be closed in 1 day.\n\nYou can add 'keep-open' label to prevent this from happening." in 20043
I0509 16:33:14.772747   20663 github.go:1365] Commenting "This PR hasn't been active in 68 days. Will be closed in 21 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 20054
I0509 16:33:15.277833   20663 github.go:1365] Commenting "This PR hasn't been active in 87 days. Will be closed in 2 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 20110
I0509 16:33:17.351331   20663 github.go:1365] Commenting "This PR hasn't been active in 82 days. Will be closed in 7 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 20203
I0509 16:33:18.555708   20663 github.go:1365] Commenting "This PR hasn't been active in 52 days. Will be closed in 37 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 20273
I0509 16:33:19.035371   20663 github.go:1365] Commenting "This PR hasn't been active in 94 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 20321
I0509 16:33:19.035402   20663 github.go:1146] Closing PR# 20321
I0509 16:33:20.272762   20663 github.go:1365] Commenting "This PR hasn't been active in 54 days. Will be closed in 35 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 20384
I0509 16:33:20.810173   20663 github.go:1365] Commenting "This PR hasn't been active in 94 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 20442
I0509 16:33:20.810220   20663 github.go:1146] Closing PR# 20442
I0509 16:33:21.282675   20663 github.go:1365] Commenting "This PR hasn't been active in 97 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 20465
I0509 16:33:21.282723   20663 github.go:1146] Closing PR# 20465
I0509 16:33:21.764517   20663 github.go:1365] Commenting "This PR hasn't been active in 94 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 20468
I0509 16:33:21.764566   20663 github.go:1146] Closing PR# 20468
I0509 16:33:23.397327   20663 github.go:1365] Commenting "This PR hasn't been active in 90 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 20666
I0509 16:33:23.397360   20663 github.go:1146] Closing PR# 20666
I0509 16:33:24.860673   20663 github.go:1365] Commenting "This PR hasn't been active in 94 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 20691
I0509 16:33:24.860706   20663 github.go:1146] Closing PR# 20691
I0509 16:33:27.198679   20663 github.go:1365] Commenting "This PR hasn't been active in 74 days. Will be closed in 15 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 20713
I0509 16:33:27.665970   20663 github.go:1365] Commenting "This PR hasn't been active in 94 days. Feel free to reopen.\n\nYou can add 'keep-open' label to prevent this from happening again." in 20749
I0509 16:33:27.666021   20663 github.go:1146] Closing PR# 20749
I0509 16:33:28.165488   20663 github.go:1365] Commenting "This PR hasn't been active in 88 days. Will be closed in 1 day.\n\nYou can add 'keep-open' label to prevent this from happening." in 20854
I0509 16:33:29.824347   20663 github.go:1365] Commenting "This PR hasn't been active in 53 days. Will be closed in 36 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 20918
I0509 16:33:32.327759   20663 github.go:1365] Commenting "This PR hasn't been active in 77 days. Will be closed in 12 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21137
I0509 16:33:33.293212   20663 github.go:1365] Commenting "This PR hasn't been active in 82 days. Will be closed in 7 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21202
I0509 16:33:35.558870   20663 github.go:1365] Commenting "This PR hasn't been active in 68 days. Will be closed in 21 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21263
I0509 16:33:36.510837   20663 github.go:1365] Commenting "This PR hasn't been active in 82 days. Will be closed in 7 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21280
I0509 16:33:36.997620   20663 github.go:1365] Commenting "This PR hasn't been active in 40 days. Will be closed in 49 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21281
I0509 16:33:38.011999   20663 github.go:1365] Commenting "This PR hasn't been active in 73 days. Will be closed in 16 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21328
I0509 16:33:38.489033   20663 github.go:1365] Commenting "This PR hasn't been active in 58 days. Will be closed in 31 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21366
I0509 16:33:39.150746   20663 github.go:1365] Commenting "This PR hasn't been active in 66 days. Will be closed in 23 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21373
I0509 16:33:42.073794   20663 github.go:1365] Commenting "This PR hasn't been active in 80 days. Will be closed in 9 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21527
I0509 16:33:42.571325   20663 github.go:1365] Commenting "This PR hasn't been active in 54 days. Will be closed in 35 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21553
I0509 16:33:46.073862   20663 github.go:1365] Commenting "This PR hasn't been active in 73 days. Will be closed in 16 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21879
I0509 16:33:46.589716   20663 github.go:1365] Commenting "This PR hasn't been active in 70 days. Will be closed in 19 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21888
I0509 16:33:47.048295   20663 github.go:1365] Commenting "This PR hasn't been active in 74 days. Will be closed in 15 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21942
I0509 16:33:48.471000   20663 github.go:1365] Commenting "This PR hasn't been active in 70 days. Will be closed in 19 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 21979
I0509 16:33:48.960820   20663 github.go:1365] Commenting "This PR hasn't been active in 70 days. Will be closed in 19 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22010
I0509 16:33:49.455406   20663 github.go:1365] Commenting "This PR hasn't been active in 57 days. Will be closed in 32 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22031
I0509 16:33:49.988932   20663 github.go:1365] Commenting "This PR hasn't been active in 46 days. Will be closed in 43 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22040
I0509 16:33:50.987792   20663 github.go:1365] Commenting "This PR hasn't been active in 70 days. Will be closed in 19 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22079
I0509 16:33:51.476127   20663 github.go:1365] Commenting "This PR hasn't been active in 49 days. Will be closed in 40 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22131
I0509 16:33:52.007854   20663 github.go:1365] Commenting "This PR hasn't been active in 48 days. Will be closed in 41 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22216
I0509 16:33:55.852716   20663 github.go:1365] Commenting "This PR hasn't been active in 67 days. Will be closed in 22 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22468
I0509 16:33:56.324712   20663 github.go:1365] Commenting "This PR hasn't been active in 62 days. Will be closed in 27 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22478
I0509 16:33:56.829328   20663 github.go:1365] Commenting "This PR hasn't been active in 39 days. Will be closed in 50 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22488
I0509 16:33:58.619141   20663 github.go:1365] Commenting "This PR hasn't been active in 65 days. Will be closed in 24 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22565
I0509 16:33:59.113020   20663 github.go:1365] Commenting "This PR hasn't been active in 31 days. Will be closed in 58 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22568
I0509 16:34:01.651126   20663 github.go:1365] Commenting "This PR hasn't been active in 61 days. Will be closed in 28 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22737
I0509 16:34:02.144003   20663 github.go:1365] Commenting "This PR hasn't been active in 33 days. Will be closed in 56 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22763
I0509 16:34:05.130096   20663 github.go:1365] Commenting "This PR hasn't been active in 59 days. Will be closed in 30 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22832
I0509 16:34:05.930335   20663 github.go:1365] Commenting "This PR hasn't been active in 37 days. Will be closed in 52 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 22857
I0509 16:34:08.083919   20663 github.go:1365] Commenting "This PR hasn't been active in 54 days. Will be closed in 35 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23074
I0509 16:34:09.077548   20663 github.go:1365] Commenting "This PR hasn't been active in 48 days. Will be closed in 41 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23095
I0509 16:34:09.545615   20663 github.go:1365] Commenting "This PR hasn't been active in 53 days. Will be closed in 36 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23119
I0509 16:34:12.072224   20663 github.go:1365] Commenting "This PR hasn't been active in 42 days. Will be closed in 47 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23254
I0509 16:34:13.572308   20663 github.go:1365] Commenting "This PR hasn't been active in 40 days. Will be closed in 49 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23306
I0509 16:34:14.689895   20663 github.go:1365] Commenting "This PR hasn't been active in 35 days. Will be closed in 54 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23344
I0509 16:34:15.150043   20663 github.go:1365] Commenting "This PR hasn't been active in 37 days. Will be closed in 52 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23362
I0509 16:34:15.632976   20663 github.go:1365] Commenting "This PR hasn't been active in 47 days. Will be closed in 42 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23382
I0509 16:34:16.152773   20663 github.go:1365] Commenting "This PR hasn't been active in 34 days. Will be closed in 55 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23400
I0509 16:34:18.335668   20663 github.go:1365] Commenting "This PR hasn't been active in 45 days. Will be closed in 44 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23489
I0509 16:34:19.379432   20663 github.go:1365] Commenting "This PR hasn't been active in 40 days. Will be closed in 49 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23498
I0509 16:34:19.859162   20663 github.go:1365] Commenting "This PR hasn't been active in 42 days. Will be closed in 47 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23502
I0509 16:34:20.402449   20663 github.go:1365] Commenting "This PR hasn't been active in 44 days. Will be closed in 45 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23513
I0509 16:34:20.942304   20663 github.go:1365] Commenting "This PR hasn't been active in 31 days. Will be closed in 58 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23521
I0509 16:34:21.410883   20663 github.go:1365] Commenting "This PR hasn't been active in 38 days. Will be closed in 51 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23565
I0509 16:34:27.391261   20663 github.go:1365] Commenting "This PR hasn't been active in 34 days. Will be closed in 55 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23733
I0509 16:34:29.446221   20663 github.go:1365] Commenting "This PR hasn't been active in 35 days. Will be closed in 54 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23775
I0509 16:34:29.935931   20663 github.go:1365] Commenting "This PR hasn't been active in 37 days. Will be closed in 52 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23785
I0509 16:34:30.408351   20663 github.go:1365] Commenting "This PR hasn't been active in 32 days. Will be closed in 57 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23788
I0509 16:34:31.816314   20663 github.go:1365] Commenting "This PR hasn't been active in 35 days. Will be closed in 54 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23810
I0509 16:34:33.268087   20663 github.go:1365] Commenting "This PR hasn't been active in 35 days. Will be closed in 54 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23830
I0509 16:34:34.239408   20663 github.go:1365] Commenting "This PR hasn't been active in 30 days. Will be closed in 59 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23846
I0509 16:34:35.400725   20663 github.go:1365] Commenting "This PR hasn't been active in 34 days. Will be closed in 55 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23866
I0509 16:34:35.885952   20663 github.go:1365] Commenting "This PR hasn't been active in 34 days. Will be closed in 55 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23867
I0509 16:34:37.362788   20663 github.go:1365] Commenting "This PR hasn't been active in 31 days. Will be closed in 58 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23919
I0509 16:34:39.675846   20663 github.go:1365] Commenting "This PR hasn't been active in 32 days. Will be closed in 57 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 23980
I0509 16:34:42.263044   20663 github.go:1365] Commenting "This PR hasn't been active in 31 days. Will be closed in 58 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 24021
I0509 16:34:45.317790   20663 github.go:1365] Commenting "This PR hasn't been active in 31 days. Will be closed in 58 days.\n\nYou can add 'keep-open' label to prevent this from happening." in 24074
```

apelisse commented:
Also, I can't track push events unfortunately. I couldn't find a way to do it by looking at Github API. So I contacted Github directly and they said it can't be done. We would have to listen to a webhook and store that locally.

eparis commented on mungegithub/mungers/close-stale-pr.go:
```diff
@@ -0,0 +1,302 @@
+	}
+}
+
+func (CloseStalePR) isStaleComment(obj *github.MungeObject, comment githubapi.IssueComment) bool {
```
This should be unit tested.

eparis commented on mungegithub/mungers/close-stale-pr.go:
```diff
@@ -0,0 +1,302 @@
+		durationToDays(closeIn)))
+}
+
+func checkAndWarn(obj *github.MungeObject, inactiveFor time.Duration, closeIn time.Duration) {
```
this 'might should be' unit tested, seems like a lot going on...

admittedly sometimes we don't test at all, and sometimes we just test `Munge`..

bgrant0607 commented:
Please see:
https://github.com/kubernetes/kubernetes/issues/24844
cc @fejta

apelisse commented:
@bgrant0607, We plan to implement these features/big picture, and this PR will certainly get refactored in the future to follow that goal. 

For now, this pull-request is mostly a way for me to get familiar with the project. 

We may decide to wait for the entire project to move forward and not merge this PR. I would be fine either way. I think it doesn't hurt to merge this (as soon as unit-tests are ready) and make smaller iterations.

bgrant0607 commented:
Please explicitly `@`-notify the author and assignee in the PR comment. I filter all github notifications that don't notify me explicitly. Additionally, PR authors who don't have write access won't be able to add the label. What would count as sufficient activity to keep the PR open for a while longer, without a label?

apelisse commented:
Yes, I will add that. Ideally, in the future, it'd be nice to notify the person that "blocks" the pull-request.

Commenting on the PR is good enough to keep it open for another 90 days.

eparis commented:
I think I'd rather not do the 'label' at all.  Just tell me people to make a comment to keep it alive...

apelisse commented:
@eparis, Yeah, it really depends on the staleness time. 

If we use 3 months, we definitely don't need the label. But @ixdy suggested closing pull-requests after 2 weeks of inactivity, and then the label would make sense.

In my opinion, we should start with three months (probably without the label) and then try to get that smaller and smaller as we can. We would literally close one third of the pull-request right now if we used 2 weeks.

I'll remove the label thing.

bgrant0607 commented:
Rather than threatening to close, I think there would be more value to aggressively prod PRs with what specifically is blocking them.

ixdy commented:
Our [documentation says](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#other-notes) we close pull requests after 2 weeks, though clearly we don't follow that (yet).

eparis commented:
@bgrant0607 agree with the long term concept.

But, threatening to close is both 'easy' from a coding automation PoV, and will hopefully cause the PR author to harass the right person to get it unstuck. It also solves the problem of actually dead PRs where no amount of automation can unstick it.

i think we need to do a better job automating moving things forward, but that doesn't make this a bad step in my mind.

bgrant0607 commented:
@eparis @ixdy I don't oppose threatening to close. I just don't think it's the first priority.

bgrant0607 commented:
What does "Can one of the admins verify this patch?" mean?

bgrant0607 commented:
Added the issue to the PR description

bgrant0607 commented:
ref #869

apelisse commented:
Should we reconsider merging this PR? (if we do, I'll rebase as needed and take another look).
While I agree with @bgrant0607 that it's not first priority. It is there, and it wouldn't hurt to merge (and some people are actually.

apelisse commented:
@eparis Can you please merge when tests are green? Thanks

bgrant0607 commented:
@apelisse Please document the behavior.

https://github.com/kubernetes/kubernetes/blob/master/docs/devel/automation.md

apelisse commented:
@bgrant0607 Excellent idea, will do in a seperate pr (obviously)

bgrant0607 commented:
Saw the doc PR. We can reevaluate the label approach later. Merging.
